### PR TITLE
Keep rust-lang/rust derived items in third_party/rust

### PR DESCRIPTION
**Pull request by blaxill**
This separates `https://github.com/rust-lang/rust/` derived items in `third_party/rust` to make their origin clear

### CODE REVIEW

**Comment by blaxill** [comment]
@daviddrysdale there is now a clean import before modification. Also I've moved `sdk/rust/oak_runtime_compat` to `oak/aslyo_rust` as
- seems more a more appropriate name&location
- my current experiments with teaclave/fortanix suggest that it makes sense to have a rust asylo layer (this)

If we decide to move to away from asylo this can be removed.

**Comment by daviddrysdale** [comment]
I think it's OK to just import a pristine copy of a *subset* of the upstream code you want to modify (that's what I [did for the proto compiler](https://github.com/project-oak/oak/commit/b3e20cbf486daa411d77329a4fbed2cef4e99c95) and for [status.proto](https://github.com/project-oak/oak/commit/84a87100fca326e24cfb60560429c27b7b5d087d)) &ndash; so I think you could squash commit 2 & 3 together.

File: third_party/rust/src/libstd/BUILD [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+load("@io_bazel_rules_rust//rust:rust.bzl", "rust_library")
+
+package(
+    licenses = ["notice"],
+)
```
Comment (daviddrysdale) [inline third_party/rust/src/libstd/BUILD]: Do we need an `exports_files(["LICENSE-APACHE", "LICENSE-MIT"])` in a `BUILD` file somewhere?

File: third_party/rust/CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,457 @@
+# Contributing to Rust
```
Comment (daviddrysdale) [inline third_party/rust/CONTRIBUTING.md]: Ding this file?

File: third_party/rust/CODE_OF_CONDUCT.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+# The Rust Code of Conduct
```
Comment (daviddrysdale) [inline third_party/rust/CODE_OF_CONDUCT.md]: Ding this file?

File: oak/asylo_rust/src/lib.rs [hunk 4]
```diff
@@ -34,13 +32,39 @@ extern crate alloc;
...
+static A: allocator::System = allocator::System;
+
+// TODO: Enclave backend that provides:
+// - Allocator
+// - Threads
+// - Mutexes
+// - TODO: Channel with automatic remote attestation
+
+/// A safe function to spawn an enclave thread. At the moment, parameters cannot
+/// be passed in this function call, unlike the rust standard library. (This can
```
Comment (daviddrysdale) [inline oak/asylo_rust/src/lib.rs]: nit: s/rust/Rust/

File: third_party/rust/README.md [hunk 5]
```diff
@@ -0,0 +1,283 @@
+# The Rust Programming Language
```
[the remark this answers is not in the document]
Comment (blaxill) [reply third_party/rust/README.md]: Done

File: third_party/rust/CONTRIBUTING.md (hunk 2 above)
Comment (blaxill) [reply third_party/rust/CONTRIBUTING.md]: Done

File: third_party/rust/CODE_OF_CONDUCT.md (hunk 3 above)
Comment (blaxill) [reply third_party/rust/CODE_OF_CONDUCT.md]: Done

File: third_party/rust/src/libstd/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,18 @@
+[package]
+name = "rust"
+version = "0.1.0"
+homepage = "https://github.com/rust-lang/rust"
+repository = "https://github.com/rust-lang/rust"
+license = "MIT/Apache-2.0"
```
[the remark this answers is not in the document]
Comment (blaxill) [reply third_party/rust/src/libstd/Cargo.toml]: Changed to "MIT OR Apache-2.0"

File: third_party/rust/src/libstd/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+description = """
+  Parts of the rust stdlib ported to asylo
```
[the remark this answers is not in the document]
Comment (blaxill) [reply third_party/rust/src/libstd/Cargo.toml]: Done

File: oak/asylo_rust/src/lib.rs (hunk 4 above)
Comment (blaxill) [reply oak/asylo_rust/src/lib.rs]: Fixed

File: third_party/rust/src/libstd/io/error.rs [hunk 8]
```diff
@@ -103,7 +88,7 @@ struct Custom {
 #[non_exhaustive]
 pub enum ErrorKind {
     /// An entity was not found, often a file.
-    NotFound,
+   NotFound,
```
Comment (daviddrysdale) [inline third_party/rust/src/libstd/io/error.rs]: Formatting looks hinky?

File: third_party/rust/src/libstd/Cargo.toml (hunk 6 above)
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply third_party/rust/src/libstd/Cargo.toml]: I'm still seeing "MIT/Apache-2.0" &ndash; lost in a merge somewhere?

File: oak/asylo_rust/src/lib.rs (hunk 4 above)
Comment (daviddrysdale) [reply oak/asylo_rust/src/lib.rs]: Not seeing this &ndash; lost in a merge somewhere?

File: third_party/rust/src/libstd/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,18 @@
+[package]
+name = "rust"
+version = "0.1.0"
+homepage = "https://github.com/rust-lang/rust"
+repository = "https://github.com/rust-lang/rust"
+license = "MIT or Apache-2.0"
```
Comment (daviddrysdale) [inline third_party/rust/src/libstd/Cargo.toml]: I believe `or` needs to be `OR`

**Comment by blaxill** [comment]
@daviddrysdale Ok should be all done including "OR". Excuse the messy thread, I'm still getting used to the github review flow.

**Pull request merged**